File: LeetCode/Study/Array&String/IntroToArray/Questions/plus-one.cpp
// https://leetcode.com/problems/plus-one/
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
  vector<int> plusOne(vector<int> &digits)
  {
    for (int i = digits.size() - 1; i >= 0; i--)
    {
      if (digits[i] < 9)
      {
        digits[i]++;
        return digits;
      }
      else
      {
        digits[i] = 0;
      }
    }

    digits.push_back(0);
    digits[0] = 1;

    return digits;
  }
};

int main()
{
  vector<int> digits{9};

  for (auto i : Solution().plusOne(digits))
  {
    cout << i << ", ";
  }
}

File: LeetCode/Easy/66-plus-one/go/plusOne.go
package plusOne

func PlusOne(digits []int) []int {
	carry := 1

	for i := len(digits) - 1; carry > 0 && i >= 0; i-- {
		digit := digits[i] + carry
		if digit == 10 {
			digits[i] = 0
		} else {
			digits[i] = digit
			carry = 0
		}
	}

	if carry == 1 {
		return append([]int{1}, digits...)
	}

	return digits
}
